### PR TITLE
Add fresnel filter for the pipeline

### PR DESCRIPTION
**Pull request by namannimmo10**

### CODE REVIEW

File: examples/fresnel-filter.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+# Load the projection data
+(   host_data,
+    host_flats,
+    host_darks,
+    angles_radians,
+    angles_total,
+    detector_y,
+    detector_x,
+) = standard_tomo(in_file, data_key, image_key, DIMENSION, PREVIEW, PAD, comm)
```
Comment (namannimmo10) [inline examples/fresnel-filter.py]:
```suggestion
host_data = \
    standard_tomo(in_file, data_key, image_key, DIMENSION, PREVIEW, PAD, comm)[0]
```

File: src/httomolib/filtering.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    pattern : str
+        Choose 'PROJECTION' for filtering projection, otherwise, will be handled
+        generically for other cases.
+
+    ratio : float
+        Control the strength of the filter. Greater is stronger.
+
+    apply_log : optional, bool
+        Apply negative log function to data being filtering.
```
Comment (namannimmo10) [inline src/httomolib/filtering.py]:
```suggestion
        Apply negative log function to data being filtered.
```

File: src/httomolib/filtering.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    # Define array to hold result. Note that, due to the padding applied, the
+    # shape of the filtered images are different to the shape of the
+    # original/unfiltered images.
+    padded_height = mat.shape[1] + pad_width*2
+    res_height = \
+        nrow if nrow < padded_height - pad_width else padded_height - pad_width
+    padded_width = mat.shape[2] + pad_width*2
+    res_width = \
+        ncol if ncol < padded_width - pad_width else padded_width - pad_width
```
Comment (namannimmo10) [inline src/httomolib/filtering.py]:
```suggestion
    res_width = min(ncol, padded_width - pad_width)
```

File: src/httomolib/filtering.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    # the number of rows in a projection image, and for sinogram images, `nrow`
+    # is the number of rows in a sinogram image).
+    (_, nrow, ncol) = mat.shape
+
+    # Define array to hold result. Note that, due to the padding applied, the
+    # shape of the filtered images are different to the shape of the
+    # original/unfiltered images.
+    padded_height = mat.shape[1] + pad_width*2
+    res_height = \
+        nrow if nrow < padded_height - pad_width else padded_height - pad_width
```
Comment (namannimmo10) [inline src/httomolib/filtering.py]:
```suggestion
    res_height = min(nrow, padded_height - pad_width)
```

**Comment by namannimmo10** [comment]
I applied the suggestions too, please feel free to merge if the tests pass for you. Cheers.

**Review by yousefmoazzam — CHANGES_REQUESTED** [review]
Unfortunately, it looks like the test for the Fresnel filter is failing on my machine for a similar reason we saw with other methods, a failed tolerance assertion:
```
========================================== FAILURES ===========================================
_____________________________________ test_fresnel_filter _____________________________________

    def test_fresnel_filter():
        in_file = 'data/tomo_standard.nxs'
        data_key = '/entry1/tomo_entry/data/data'
        image_key = '/entry1/tomo_entry/data/image_key'
        dimension = 1
        preview = [None, None, None]
        pad = 0

        host_data = \
            standard_tomo(in_file, data_key, image_key, dimension, preview, pad, comm)[0]

        data = cp.array(host_data)

        pattern = 'PROJECTION'
        ratio = 100.0
        filtered_data = fresnel_filter(data, pattern, ratio)

        for _ in range(5):
            assert_allclose(cp.mean(filtered_data), 802.1125)
            assert_allclose(cp.max(filtered_data), 1039.5293)
            assert_allclose(cp.min(filtered_data), 95.74562)

        pattern = 'SINOGRAM'
        filtered_data = fresnel_filter(data, pattern, ratio)
        for _ in range(5):
>           assert_allclose(cp.mean(filtered_data), 806.74347)

tests/test_filtering.py:36:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

actual = array(806.74335, dtype=float32), desired = 806.74347, rtol = 1e-07, atol = 0
err_msg = '', verbose = True

    def assert_allclose(actual, desired, rtol=1e-7, atol=0, err_msg='',
                        verbose=True):
        """Raises an AssertionError if objects are not equal up to desired tolerance.

        Args:
             actual(numpy.ndarray or cupy.ndarray): The actual object to check.
             desired(numpy.ndarray or cupy.ndarray): The desired, expected object.
             rtol(float): Relative tolerance.
             atol(float): Absolute tolerance.
             err_msg(str): The error message to be printed in case of failure.
             verbose(bool): If ``True``, the conflicting
                 values are appended to the error message.

        .. seealso:: :func:`numpy.testing.assert_allclose`

        """  # NOQA
>       numpy.testing.assert_allclose(
            cupy.asnumpy(actual), cupy.asnumpy(desired),
            rtol=rtol, atol=atol, err_msg=err_msg, verbose=verbose)
E       AssertionError:
E       Not equal to tolerance rtol=1e-07, atol=0
E
E       Mismatched elements: 1 / 1 (100%)
E       Max absolute difference: 0.00012283
E       Max relative difference: 1.52256617e-07
E        x: array(806.74335, dtype=float32)
E        y: array(806.74347)

/dls/science/users/twi18192/conda-envs/httomolib/lib/python3.9/site-packages/cupy/testing/_array.py:24: AssertionError
------------------------------------ Captured stdout call -------------------------------------
The full dataset shape is (220, 128, 160)
[0] Pad values are (0, 0).
data/tomo_standard.nxs
/entry1/tomo_entry/data/data
0:180, :, :
[0] Data shape is (180, 128, 160) of type uint16
=================================== short test summary info ===================================
FAILED tests/test_filtering.py::test_fresnel_filter - AssertionError:
1 failed, 4 passed in 32.46s
```

I'm happy to make the necessary changes and add them to the PR since it's my machine that's seeing the error.

**Comment by yousefmoazzam** [comment]
In the end it was just one calculation in the Fresnel filter test that was failing the tolerance assertion, let me know if that test still passes for you with the change I've made.

**Comment by namannimmo10** [comment]
Thank you for fixing the test! It works for me, too.

**Pull request merged**